bokeh/server/session.py

```diff
@@ -179,26 +180,6 @@ def _document_changed(self, event):
         may_suppress = self._current_patch is not None and \
                        self._current_patch.should_suppress_on_change(event)
 
-        if isinstance(event, SessionCallbackAdded):
-            if isinstance(event.callback, PeriodicCallback):
-                self._add_periodic_callback(event.callback)
-            elif isinstance(event.callback, TimeoutCallback):
-                self._add_timeout_callback(event.callback)
-            else:
-                raise ValueError("Expected callback of type PeriodicCallback or TimeoutCallback, got: %s" % event.callback)
-
-            return
-
-        elif isinstance(event, SessionCallbackRemoved):
-            if isinstance(event.callback, PeriodicCallback):
-                self._remove_periodic_callback(event.callback)
-            elif isinstance(event.callback, TimeoutCallback):
-                self._remove_timeout_callback(event.callback)
-            else:
-                raise ValueError("Expected callback of type PeriodicCallback or TimeoutCallback, got: %s" % event.callback)
-
-            return
-
         if self._pending_writes is None:
             raise RuntimeError("_pending_writes should be non-None when we have a document lock, and we should have the lock when the document changes")
 
```

Comment:
this is not satisfied for add/remove SessionCallback Events.. AFAIK we don't need to lock the document here so just decorate `_add_periodic_callback` or `_remove_periodic_callback` with `_needs_document_lock` is not really necessary (would it even work since `_document_changed` is called before these?).
We could eventually only call `_add_periodic_callback` or `_remove_periodic_callback` on the dispatch.... 

I may be missing something.. @havocp do you have thoughts?